Add the new Audience_Settings class and datapoints.

ankitrox opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #10153 

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.
- [x] Ensure there are no unexpected significant changes to file sizes.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

techanvil commented on includes/Modules/Analytics_4/Audience_Settings.php:
```diff
@@ -0,0 +1,142 @@
+	 *
+	 * @return array The sanitized settings.
+	 */
+	private function sanitize( $settings, $existing_settings = null ) {
+		$option = $existing_settings ?? $this->get_default();
```
As we only need to call this method in `get_sanitize_callback()` it doesn't need the second argument.

Also, seeing as it's only being called there and we are passing `$option` to it, let's rename this argument and variable to make it more clear. We should also preserve current settings by calling `get()` instead of `get_default()`:

```suggestion
	private function sanitize( $option ) {
		$new_option = $this->get();
```

techanvil commented on tests/phpunit/integration/Modules/Analytics_4/Audience_SettingsTest.php:
```diff
@@ -0,0 +1,191 @@
+		);
+	}
+
+	/**
+	 * @group 10up
+	 */
```
Please remove this incorrect group.

```suggestion
```

techanvil commented on tests/phpunit/integration/Modules/Analytics_4/Audience_SettingsTest.php:
```diff
@@ -0,0 +1,191 @@
+
+		$this->options           = new Options( new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE ) );
+		$this->audience_settings = new Audience_Settings( $this->options );
+		$this->user_id           = $this->factory()->user->create( array( 'role' => 'administrator' ) );
+
+		wp_set_current_user( $this->user_id );
```
This appears to be unnecessary:

```suggestion
```

techanvil reviewed (changes requested):
Thanks @ankitrox, this is most of the way there but needs a few tweaks. Please see my comments.

ankitrox commented:
Thanks @techanvil for reviewing the PR and adding the feedback.

I have addressed the feedback and pushed the changes.

Assigning this to you for another review.

Thanks

techanvil commented on includes/Modules/Analytics_4.php:
```diff
@@ -1469,6 +1491,44 @@ protected function create_data_request( Data_Request $data ) {
+						$new_settings['audienceSegmentationSetupCompletedBy'] = $data['audienceSegmentationSetupCompletedBy'];
+					}
+
+					if ( isset( $data['availableAudiences'] ) && is_array( $data['availableAudiences'] ) ) {
+						$new_settings['availableAudiences'] = $data['availableAudiences'];
+					}
+
+					if ( isset( $data['availableAudiencesLastSyncedAt'] ) ) {
+						$new_settings['availableAudiencesLastSyncedAt'] = intval( $data['availableAudiencesLastSyncedAt'] );
+					}
+
```
As per my previous comment, we shouldn't handle settings other than `audienceSegmentationSetupCompletedBy` in the `POST:save-audience-settings` datapoint.

> we don't want to set `audienceSegmentationSetupCompletedBy` to the current user ID here, rather the idea is we should allow the setting to be passed in, but not handle any other settings for now

I realise my suggested edit didn't make this entirely clear. I'll apply this change myself to save the round trip.

```suggestion
```

techanvil reviewed (changes requested):
LGTM, thanks @ankitrox!

Note that I've confirmed the failing E2E tests are unrelated to the changes in this PR.
